--- a/Makefile
+++ b/Makefile
@@ -1,23 +1,27 @@
 #Makefile
 #char code "LF(UNIX)" is required!!
 
-OPTIONS = -O1
-PROGS = main
+MAKEFLAGS += --no-builtin-rules
 
-all: $(PROGS)
+CC ?= cc
+CFLAGS ?= -O1
+CFLAGS += -D_GNU_SOURCE
+LDFLAGS ?=
+LDFLAGS += -l rt
+
+SRC = main.c encapsulate.c session.c nat.c
+DEPS = $(patsubst %.c, %.h, $(SRC))
+OBJ = $(patsubst %.c, %.o, $(SRC))
+
+.PHONY: all
+all: map-e
+
+.PHONY: clean
+clean:
+	rm -f map-e *.o
 
-main: main.o encapsulate.o session.o nat.o
-	cc $(OPTIONS) -o map-e main.o encapsulate.o session.o nat.o -l rt
-	rm *.o
+map-e: $(OBJ)
+	$(CC) $(LDFLAGS) -o $@ $^
 
-main.o : main.c
-	cc $(OPTIONS) -c main.c
-
-encapsulate.o : encapsulate.c
-	cc $(OPTIONS) -c encapsulate.c
-
-session.o : session.c
-	cc $(OPTIONS) -c session.c
-
-nat.o : nat.c
-	cc $(OPTIONS) -c nat.c
+$(OBJ): %.o: %.c $(DEPS)
+	$(CC) -c $(CFLAGS) -o $@ $<
